src/test/java/org/opensearch/security/dlic/rest/api/InternalUsersApiActionValidationTest.java

```diff
@@ -94,6 +136,48 @@ public void validateAndUpdatePassword() throws Exception {
         assertEquals(RestStatus.INTERNAL_SERVER_ERROR, result.status());
     }
 
+    @Test
+    public void validateSecurityRolesWithMutableRolesMappingConfig() throws Exception {
+        final var internalUsersApiAction = createInternalUsersApiAction();
+
+        // should ok to set regular role with mutable role mapping
+        var userJson = objectMapper.createObjectNode().set("opendistro_security_roles", objectMapper.createArrayNode().add("regular_role"));
+        var result = internalUsersApiAction.validateSecurityRoles(SecurityConfiguration.of(userJson, "some_user", configuration));
+        assertTrue(result.isValid());
+        // should be ok to set reserved role with mutable role mapping
+        userJson = objectMapper.createObjectNode().set("opendistro_security_roles", objectMapper.createArrayNode().add("kibana_read_only"));
+        result = internalUsersApiAction.validateSecurityRoles(SecurityConfiguration.of(userJson, "some_user", configuration));
+        assertTrue(result.isValid());
+        // should be ok to set static role with mutable role mapping
+        userJson = objectMapper.createObjectNode().set("opendistro_security_roles", objectMapper.createArrayNode().add("all_access"));
+        result = internalUsersApiAction.validateSecurityRoles(SecurityConfiguration.of(userJson, "some_user", configuration));
+        assertTrue(result.isValid());
+        // should not be ok to set hidden role with mutable role mapping
+        userJson = objectMapper.createObjectNode().set("opendistro_security_roles", objectMapper.createArrayNode().add("some_hidden_role"));
+        result = internalUsersApiAction.validateSecurityRoles(SecurityConfiguration.of(userJson, "some_user", configuration));
+        assertFalse(result.isValid());
+    }
+
+    @Test
+    public void validateSecurityRolesWithImmutableRolesMappingConfig() throws Exception {
```

Comment:
Its really confusing me that there are 2 ways that roles mapping can be done to map a user directly to a role.

1. Modify the `opendistro_security_roles` attribute of a user - this modifies the `internaluser` config entry and not the `roles_mapping` config entry
2. Update the `roles_mapping` and add the `username` in the list of `users`. 

I had to spend a little time digesting that this is simulating the behavior of a `PUT _plugins/_security/api/internalusers/<username>` call where the config entries for the role and roles_mapping already exist. In this case in the test setup it creates 4 roles where the mapping will fail:

1. `some_hidden_role` - The role is hidden
2. `some_role_with_hidden_mapping` - The role is fine, but the roles mapping is hidden
3. `some_role_with_reserved_mapping` - The role is fine, but the roles mapping is reserved
4. `some_role_with_static_mapping` - The role is fine, but the roles mapping is static

Reply:
Not exactly but closer. In any case a tests on PATCH should be added.